text/0008-ingest-gem-content.md

```diff
@@ -0,0 +1,332 @@
+If we had to consider every file as a possible orphan, we would have to search the entire file table without constraints for a matching checksum before deleting a file.
+If we instead constrain our deduplication to only files within the same gem, we get most of the benefit without the increase in file overlaps.
+When only files within the same gem are considered for deduplication, we may miss a few small optimizations, but we reduce complexity.
+Since the files within a gem can only change with versions of that gem, there's less likelihood for race conditions.
+If one gem adds a file, while another yanks the only other reference to that file, we could end up with the incorrect state.
+This is a much smaller problem within the same gem where we already assume consistency between versions and do not have to consider concurrent uploads of any other gems.
+
+Supporting this approach is npmjs.com, which uses package name and checksum to fetch file content.
+Since we have no information about how much space it would save, we also can't appropriately measure the benefit of global hashing.
+
+Once enough gems are indexed, we could estimate the total savings of this approach and consider whether it makes sense to change.
+Starting with the solution as proposed will reduce the complexity of changing.
+Moving from global index to gem-scoped index is more complicated than gem-scoped to global.
+The solution we propose in this RFC could be converted to hash globally with a few simple aws cli commands.
+The reverse, global to gem-scoped migration, would require re-parsing every gem manifest to copy files to the appropriate gem directory.
+
+### Binary Files
+
+There are many ways to detect binary files and choices to be made for serving them.
+
+In the interest of providing gem contents for the purpose of browsing on the web, we think it is best not to serve the binary portions of a gem.
+Binary files are not able to be browsed or diff'd in any meaningful way on a web page.
+
+Avoiding binaries solves many of the problems listed in drawbacks and reduces storage size significantly.
+For example, the `libv8-node-18.13.0.0-arm64-darwin.gem` we examined is 30MB.
+Uncompressed, the whole gem is about 98MB, while the single binary `.a` file is 97MB.
+Not storing the only binary file in this gem saves 99% of the uncompressed storage space while providing almost all the benefit of offering gem browsing.
+
+A user that wants to examine a binary file in a gem is effectively downloading more than the file size of the gem.
+They would be better served downloading the gem for that purpose.
+
+#### Detecting Binary Files
+
+There are many solutions that can be found for detecting binary files.
+We intend at this time to use `grep` to quickly print a list of files that it thinks are binary, and therefore not text searchable.
+We do this by relying on `grep` to output when a binary file matches.
```

Comment:
I think this might be significantly slower than using https://rubygems.org/gems/ruby-magic, since it's spawning a subprocess?

Reply:
I don't have an estimate of timing. The benefit of grep is that it would return all binary files in the response of one subprocess. However, I still won't have content-type which may mean using `ruby-magic` rather than `file`. As I started prototyping this to understand it better, I ran into the problem where checksum, file type, and `binary?` all needed to load the file into memory in a different way. Maybe I could put in here all the options we can think of, and then leave it as an open question to benchmark the different approaches. `ruby-magic` is on my radar to try, in particular because `file` may vary in output between *nix systems.